### PR TITLE
[Bug fix] Fix non-finite prod_bw gradients for zero inputs

### LINKED ISSUE #54551: [Bounty $1,000] ttnn.prod_bw returns non-finite gradients for zero inputs
## Summary

`ttnn.prod_bw` computes the gradient as

```text
prod(input) * grad / input
```

which is `reciprocal(input)` multiplied by a broadcast `prod(input) * grad`. Where the input
contains an exact zero, `reciprocal(0)` is `+inf` and the surviving product is `0`, so the
gradient at that position is non-finite instead of the finite value autograd defines.

```text
input:    [2, 0, 4]
grad:     1
expected: [0, 8, 0]
current:  [0, non-finite, 0]
```

This is a source-level analysis: I do not have a card, so the host replay below renders the
device `inf * 0` as a non-finite value without asserting which one appears in silicon. The defect
is that the gradient is non-finite where the correct value is finite.

## Root cause

Both branches multiply by `reciprocal(input)`:

- `dim=None`:
  [`unary_backward.cpp#L1668-L1676`](https://github.com/tenstorrent/tt-metal/blob/a5cd86212e/ttnn/cpp/ttnn/operations/eltwise/unary_backward/unary_backward.cpp#L1668-L1676)
  — `multiply(reciprocal(input), fill_first_val_into_tensor(prod * grad))`.
- per-dimension: `reciprocal_input` is built once at
  [`L1711`](https://github.com/tenstorrent/tt-metal/blob/a5cd86212e/ttnn/cpp/ttnn/operations/eltwise/unary_backward/unary_backward.cpp#L1711)
  and feeds every `dim` branch, e.g. `dim=-1` at
  [`L1722`](https://github.com/tenstorrent/tt-metal/blob/a5cd86212e/ttnn/cpp/ttnn/operations/eltwise/unary_backward/unary_backward.cpp#L1722).
  All four dimensions are affected, not only `dim=-1`.

At a zero input the product of the other values has already been lost in the reduction, so it
cannot be recovered by the division. The formula is valid only when the reduced input has no zeros.

## Reproducer

Host only, no device required (`pip install torch`):

```python
"""Source-level replay of ttnn.prod_bw against torch.prod autograd.

Replays the composite expression from tt-metal main @ a5cd86212e. Runs on host only:
it does not need a Tenstorrent device.
"""

import math
import torch


def source_prod_bw(values, grad):
    """unary_backward.cpp L1668-1676: reciprocal(input) * broadcast(prod(input) * grad)."""
    product = math.prod(values)
    out = []
    for v in values:
        if v == 0.0:
            # reciprocal(0) is +inf on device; inf * 0 is non-finite
            out.append(float("inf") * (product * grad))
        else:
            out.append(product * grad / v)
    return out


def source_prod_bw_dim(rows, grad):
    """unary_backward.cpp L1711-1722, dim=-1: bcast(reciprocal(input), prod * grad, MUL)."""
    return [source_prod_bw(row, grad) for row in rows]


def torch_prod_bw(values, grad):
    t = torch.tensor(values, dtype=torch.float64, requires_grad=True)
    torch.prod(t).backward(torch.tensor(grad, dtype=torch.float64))
    return t.grad.tolist()


def torch_prod_bw_dim(rows, grad):
    t = torch.tensor(rows, dtype=torch.float64, requires_grad=True)
    torch.prod(t, dim=-1, keepdim=True).backward(
        torch.full((len(rows), 1), grad, dtype=torch.float64))
    return t.grad.tolist()


def show(v):
    return "non-finite" if isinstance(v, float) and not math.isfinite(v) else f"{v:g}"


print("dim=None")
for values in ([2.0, 3.0, 4.0], [2.0, 0.0, 4.0], [0.0, 0.0, 4.0], [1.0, 5.0, 0.0]):
    actual, expected = source_prod_bw(values, 1.0), torch_prod_bw(values, 1.0)
    ok = all(a == e for a, e in zip(actual, expected))
    print(f"  input={values}  expected={expected}  "
          f"current=[{', '.join(show(a) for a in actual)}]  {'ok' if ok else 'DIVERGES'}")

print("dim=-1")
for rows in ([[2.0, 3.0, 4.0], [1.0, 5.0, 3.0]], [[2.0, 0.0, 4.0], [1.0, 5.0, 3.0]]):
    actual, expected = source_prod_bw_dim(rows, 1.0), torch_prod_bw_dim(rows, 1.0)
    ok = all(a == e for ar, er in zip(actual, expected) for a, e in zip(ar, er))
    print(f"  input={rows}  expected={expected}  "
          f"current=[{', '.join('[' + ', '.join(show(a) for a in r) + ']' for r in actual)}]  "
          f"{'ok' if ok else 'DIVERGES'}")
```

Output:

```text
dim=None
  input=[2.0, 3.0, 4.0]  expected=[12.0, 8.0, 6.0]  current=[12, 8, 6]  ok
  input=[2.0, 0.0, 4.0]  expected=[0.0, 8.0, 0.0]  current=[0, non-finite, 0]  DIVERGES
  input=[0.0, 0.0, 4.0]  expected=[0.0, 0.0, 0.0]  current=[non-finite, non-finite, 0]  DIVERGES
  input=[1.0, 5.0, 0.0]  expected=[0.0, 0.0, 5.0]  current=[0, 0, non-finite]  DIVERGES
dim=-1
  input=[[2.0, 3.0, 4.0], [1.0, 5.0, 3.0]]  expected=[[12.0, 8.0, 6.0], [15.0, 3.0, 5.0]]  current=[[12, 8, 6], [15, 3, 5]]  ok
  input=[[2.0, 0.0, 4.0], [1.0, 5.0, 3.0]]  expected=[[0.0, 8.0, 0.0], [15.0, 3.0, 5.0]]  current=[[0, non-finite, 0], [15, 3, 5]]  DIVERGES
```

## Expected behaviour

The gradient of a product with respect to one factor is the product of the *other* factors, which
is always finite for finite inputs:

- no zero in the reduced input: the current formula is correct;
- exactly one zero: zero everywhere except at that position, which gets `grad` times the product
  of the non-zero values;
- two or more zeros: zero everywhere.

Suggested tests: exact zeros in both paths, one zero and several zeros, and a non-finite check
before PCC — a PCC comparison alone can pass or silently degrade on a non-finite element.

### PR DESCRIPTION
**Pull request by kanapitsas**
### PR Category

Bug fix

### Summary

Fixes #54551.

`ttnn.prod_bw` computed the gradient as `grad * prod(input) / input`, which is non-finite
when a reduced slice contains zero.

The change replaces zeros with one during the reduction, counts them, and selects the
per-element factor:

```text
0 zeros  -> 1 / input          (the previous formula)
1 zero   -> 1 at the zero, 0 elsewhere
2+ zeros -> 0
```

The existing `grad * prod` broadcast remains unchanged.

### Tests

- Added coverage for `dim=None`, every supported dimension, and 0–3 zeros per slice.
- Tests check finiteness and compare with PyTorch autograd. Expected zeros require `atol=0`.
- Host model: 19,644 cases, 0 mismatches.

### Device reproducer

On `main` the printed count is non-zero; with this change it is `0` and the value at the
zero is `1.0`.

```python
import torch
import ttnn

device = ttnn.open_device(device_id=0)

x = torch.ones((1, 1, 32, 32), dtype=torch.bfloat16)
x[0, 0, 0, 1] = 0.0                      # one zero: gradient is 1 there, 0 elsewhere
grad = torch.zeros((1, 1, 32, 32), dtype=torch.bfloat16)
grad[0, 0, 0, 0] = 1.0                   # dim=None grad is read from position 0

tt_x = ttnn.from_torch(x, layout=ttnn.TILE_LAYOUT, device=device)
tt_grad = ttnn.from_torch(grad, layout=ttnn.TILE_LAYOUT, device=device)

out = ttnn.to_torch(ttnn.prod_bw(tt_grad, tt_x)[0]).float()
print("non-finite gradient values:", (~torch.isfinite(out)).sum().item())
print("value at the zero:", out[0, 0, 0, 1].item(), "(expected 1.0)")

ttnn.close_device(device)
```

### Notes for reviewers

- Reduced selectors use `keepdim=true` and `binary_ng` broadcasting.
- `unary_backward.cpp` compiles with this change (g++-12, aarch64 Linux, the project's
  `-Wall -Werror`), checked on a deliberately broken variant as well. I have no Tenstorrent
  hardware: the tree is not linked or run on device locally, so this PR's CI is the first
  device run of the new test.

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1218118069917566

### CODE REVIEW

File: tests/ttnn/nightly/unit_tests/operations/eltwise/backward/test_backward_prod.py [hunk 1]
```diff
@@ -88,3 +88,55 @@ def test_bw_prod_default_both(input_shapes, device):
...
+    [-4, -3, -2, -1, 0, 1, 2, 3, None],
+)
+@pytest.mark.parametrize("num_zeros", [0, 1, 2, 3], ids=["no_zero", "one_zero", "two_zeros", "three_zeros"])
+def test_bw_prod_with_zeros(input_shapes, dim, num_zeros, device):
+    all_dimensions = dim is None
+
+    torch.manual_seed(0)
+    # +/-1 keeps the full product exact. The zeros inserted below are the only zeros.
+    sign = torch.randint(0, 2, input_shapes).float() * 2 - 1
+    in_data = sign.bfloat16()
```
[the remark this answers is not in the document]
Comment (kanapitsas) [reply tests/ttnn/nightly/unit_tests/operations/eltwise/backward/test_backward_prod.py]: You are right, and it is worse than it looks: I replayed the shipped test against a bfloat16 model of the patched op, standing in a deliberately wrong implementation that keeps only the sign of `product(other factors)`. It passes all 64 non-skipped cases of `test_bw_prod_with_zeros`.

The reason that matrix is +/-1 is dynamic range: with `dim=None` the product runs over up to 12288 factors, so any magnitude other than 1 leaves the bfloat16 range. I kept it for the zero-count logic and added `test_bw_prod_zero_keeps_other_factors`, which reduces along a single dim over exact powers of two. The product along `dim` and every `product/x` quotient stay exact in bfloat16, and the gradient at a lone zero comes out at 2, 16, up to 2^17 depending on the dim, instead of +/-1.

Same replay on the new test: the sign-only implementation fails 16 of 24 cases. The 8 it passes are the `two_zeros` ones, where the true gradient is zero everywhere and no test can discriminate. The patched path passes 24/24, worst relative error 0.78% against a pessimistic SFPU-reciprocal model, under the 2% tolerance, with no violation of `atol=0` where the gradient must be exactly zero.

File: tests/ttnn/nightly/unit_tests/operations/eltwise/backward/test_backward_prod.py [hunk 2]
```diff
@@ -88,3 +88,55 @@ def test_bw_prod_default_both(input_shapes, device):
...
+    if all_dimensions:
+        in_data.view(-1)[:num_zeros] = 0.0
+    else:
+        if input_shapes[dim] < num_zeros:
+            pytest.skip(f"dim {dim} of {list(input_shapes)} is too short for {num_zeros} zeros")
+        in_data.index_fill_(dim, torch.arange(num_zeros), 0.0)
+    in_data = in_data.detach().clone().requires_grad_(True)
+
+    input_tensor = ttnn.Tensor(in_data.detach().bfloat16(), ttnn.bfloat16).to(ttnn.TILE_LAYOUT).to(device)
```
[the remark this answers is not in the document]
Comment (kanapitsas) [reply tests/ttnn/nightly/unit_tests/operations/eltwise/backward/test_backward_prod.py]: Done: `test_bw_prod_with_zeros` is now parameterized over TILE and ROW_MAJOR input layouts, covering `dim=None` and every supported dim through the new selector.

One reservation worth flagging, since this is not exercised by the PR's CI. With a ROW_MAJOR input the selector multiplies operands of mismatched layouts: `reciprocal(safe_input)` and `is_zero` stay ROW_MAJOR, while `zero_count` is explicitly tiled, so `no_zero` and `one_zero` are TILE. `should_use_row_major_path` (`binary_ng_program_factory.cpp:131-140`) requires a, b and the output to all be ROW_MAJOR, so those two multiplies take the tiled program factory with a row-major operand, and nothing tilizes it on the way in. Separately, this op's own sweep still invalidates every ROW_MAJOR vector (`tests/sweep_framework/sweeps/reduction/backward/prod_bw/prod_bw.py:76`), though the reason it gives no longer matches any guard in the tree.

So if the nightly disagrees with the `row_major` vectors, that mismatched multiply is where I would look first.

**Comment by jberkowitzTT** [comment]
@kanapitsas - Thanks for adding the non-unit-product coverage and parameterizing the tests for ROW_MAJOR input.

Before this satisfies the bounty, please:
- Run the new ROW_MAJOR cases on supported hardware/CI and attach a passing result.
- Remove or update the stale prod_bw sweep invalidation that currently skips ROW_MAJOR vectors.

**Comment by kanapitsas** [comment]
Hi @jberkowitzTT both requested changes are pushed. Could you approve the PR workflows? I'll follow up with the ROW_MAJOR test result once they finish.

**Comment by kanapitsas** [comment]
Thanks for running the workflows. Both requests are now addressed.

- Hardware result: [ttnn eltwise group 3 on wh_n300_civ2](https://github.com/tenstorrent/tt-metal/actions/runs/34632312904/job/103427722017) ran `test_prod_bw_zero.py` with 152 passed, 16 skipped, and 0 failed, including 64 passing ROW_MAJOR cases. The two simulator jobs produced the same result. The skips are expected for axes that are too short to place two or three zeros.
- Sweep invalidation: it now checks `keepdim` and the `dim` type first. Three further rules apply to ROW_MAJOR vectors only, so TILE generates exactly as before: `dim` has to address an axis that exists, the dtype has to be the documented BFLOAT16, and both the input and the gradient have to meet the ROW_MAJOR row width. The generator now has 11,264 valid vectors, up from 10,560; all 704 new vectors are ROW_MAJOR and build a valid Torch golden.

The PR's only failing job is unrelated to this change: the Blackhole runtime simulator hits `tensix_stallwait: wait_res=0x20`. The [same failure](https://github.com/tenstorrent/tt-metal/actions/runs/34641629798) also occurs on an unrelated branch.

I also found a pre-existing issue while checking the sweep: with both the input and gradient in ROW_MAJOR, `prod_bw(dim=2)` returns incorrect values on the Blackhole simulator, measured against main at a3ece1ae1e without this PR. On a `(2, 2, 32, 32)` input of signed powers of two with no zeros, the first four results are `[-2^35, 2^35, 2^35, -2^35]`, while Torch returns `[-254976, -132096, 13184, 20224]`. The PR tests use a TILE gradient and pass all dimensions. I have not tested this case on silicon. I can report it separately if useful, and I can also move the tests back under nightly if you prefer them there.

**Comment by jberkowitzTT** [comment]
@kanapitsas - in reviewing this PR, we've determined that this satisfies the bounty requirements. Thanks for the great work!

I'll note the standard next steps in the main issue

**Review by bbradelTT — APPROVED** [review]

**Review by mateusznowakTT — APPROVED** [review]

File: ttnn/cpp/ttnn/operations/eltwise/unary_backward/unary_backward.cpp [hunk 3]
```diff
@@ -1663,19 +1665,34 @@ std::vector<Tensor> prod_bw(
...
+        zero_count = ttnn::operations::unary_backward::change_layout_to_tile(zero_count, output_memory_config);
+    }
+    Tensor no_zero = ttnn::eqz(zero_count, output_memory_config);
+    Tensor one_zero = ttnn::eq_unary(zero_count, 1.0f, output_memory_config);
+    Tensor per_element = ttnn::add(
+        ttnn::multiply(ttnn::reciprocal(safe_input, output_memory_config), no_zero, std::nullopt, output_memory_config),
+        ttnn::multiply(is_zero, one_zero, std::nullopt, output_memory_config),
+        std::nullopt,
+        output_memory_config);
+
```
Comment (fplavecTT) [inline ttnn/cpp/ttnn/operations/eltwise/unary_backward/unary_backward.cpp:1689]: Temporary tensors can be deallocated after their last use to reduce peak memory usage. For example, `safe_input`, `zero_count` and `is_zero` are not used beyond this point, so they can be deallocated here. If my other idea above is implemented, then `zero_count` and `is_zero` will have been deallocated even earlier, but `safe_input` and the new tensor I suggested creating can be deallocated here.

File: ttnn/cpp/ttnn/operations/eltwise/unary_backward/unary_backward.cpp [hunk 4]
```diff
@@ -1663,19 +1665,34 @@ std::vector<Tensor> prod_bw(
...
     }
 
+    // Keep the count broadcastable with the input.
+    const std::optional<std::variant<int, int64_t, ttsl::SmallVector<int>>> reduce_dim = dim;
+    Tensor zero_count = ttnn::sum(is_zero, reduce_dim, /*keepdim=*/true, output_memory_config);
+    if (zero_count.layout() == Layout::ROW_MAJOR && zero_count.storage_type() == StorageType::DEVICE) {
+        zero_count = ttnn::operations::unary_backward::change_layout_to_tile(zero_count, output_memory_config);
+    }
+    Tensor no_zero = ttnn::eqz(zero_count, output_memory_config);
+    Tensor one_zero = ttnn::eq_unary(zero_count, 1.0f, output_memory_config);
```
Comment (fplavecTT) [inline ttnn/cpp/ttnn/operations/eltwise/unary_backward/unary_backward.cpp:1683]: Rather than calculating `no_zero` and `one_zero` separately, they could be combined into a new tensor by comparing `is_zero` and `zero_count` for equality (and `zero_count` and `is_zero` can then be deallocated immediately after doing this). This handles all cases at once: a non-zero element survives only where the count is 0, and a zero survives only where it is the only zero in the reduced slice.
This new tensor can then be multiplied by reciprocal of `safe_input` to obtain `per_element`.

File: ttnn/cpp/ttnn/operations/eltwise/unary_backward/unary_backward.cpp [hunk 5]
```diff
@@ -1663,19 +1665,34 @@ std::vector<Tensor> prod_bw(
...
+    Tensor per_element = ttnn::add(
+        ttnn::multiply(ttnn::reciprocal(safe_input, output_memory_config), no_zero, std::nullopt, output_memory_config),
```
Comment (fplavecTT) [inline ttnn/cpp/ttnn/operations/eltwise/unary_backward/unary_backward.cpp:1685]: As suggested above, this can then be replaced with multiplication by a reciprocal of `safe_input`. As a further optimization, to multiply by reciprocal, one can avoid a separate op for reciprocal and fuse it with multiply, which accepts `lhs_activations`/`rhs_activations`.
See ttnn/cpp/ttnn/operations/experimental/deepseek_prefill/routed_expert_ffn/routed_expert_ffn_bh.cpp lines 111-123 for an example of this. The main difference is that here we would use `UnaryOpType::RECIP` instead of `UnaryOpType::SILU`.

**Review by fplavecTT — COMMENTED** [review]
Good change, but it could be made more efficient via a simple refactor with fewer temporary tensors, folding reciprocal into multiply, and by deallocating temporary tensors after their last use (details below).
